docs/internal/foreign_code.md

```diff
@@ -62,32 +65,210 @@ We hope that the region abstraction will make programming with sandboxes natural
+
+Mapping C++ constructs to Verona
+--------------------------------
+
+Two things are explicitly out of scope from Verona's C++ interop layer:
+
+ - Subclassing a C++ class in Verona.
+ - Subclassing a Verona class in C++.
+
+It *is* possible to invoke callbacks from C++ that invoke Verona code and so it is possible to create, on the C++ side of the interop layer, a C++ class that subclasses another C++ class and invokes Verona callbacks for all (or some) of its methods.
+Most of the rest of C++ should be possible to expose as Verona concepts.
+
+C++ namespaces
+--------------
+
+C++ namespaces are singleton nested classes within the class exposed for the library.
+Everything within the namespace is exposed within this context in the same way that it would be if it were declared in the top-level class.
+For example:
+
+```c++
+// In libfoo
+namespace SomeStuff
+{
+  int x;
+};
+```
+
+Imported into Verona looks like this:
 
 ```verona
-class LLVM
+import "libfoo" as F;
+
+var f = F();
+var x = f.SomeStuff.x : F.int;
+```
+
+### Classes
+
+C++ classes have a number of aspects that make them more complex than C structures.
+
+*Visibility specifiers* are respected by Verona code and there is no way to make a Verona class or function a `friend` of a C++ class.
+If you want to access a `private` field of a C++ object then this should be done by exposing a `friend` function on the C++ side of the interop layer.
+
+Inheritance is represented in Verona as each subclass defining a new concrete type.
+As with normal Verona code, there is *no* implicit conversion from a pointer to a C++ class to a pointer to any of its superclasses, C++ structures passed by pointer to C++ functions must be explicitly converted.
+This is one of the places where a programmer writing a wrapper library is likely to want to write some code to simplify the interface (this is not needed for correctness, only for the code to be user friendly).
+
+Verona's unified call syntax provides a mechanism for disambiguating fields in C++ classes that are in both a base class and a subclass.
+For example, consider this C++ code:
+
+```c++
+struct Super
 {
-	[#native, #interface_language="c++"]
-	imports library("libLLVM.so", "llvm.h");
-}
+  void x();
+};
+struct Sub : public Super
+{
+  void x();
+};
+```
+
+In Verona, given an instance of `Sub`, you could call `Super::x()` like this:
+
+```verona
+var s = CXXLib.Sub();
+CXXLib.Super.x(CXXLib.static_cast[CXXLib.Sub](s));
+// The following two are equivalent 
+s.x();
+CXXLib.Sub.x(s);
 ```
 
-This includes two attributes that tell the compiler the format of the library (native code) and how to parse the header (which, in this example, would have to be something provided by the user that included the desired headers).
-Once C++ module support is more widespread, module imports can take the place of header inclusions.
+### Templates
```

Comment:
I think we also need to make a strong restriction here:
* Using Verona types in C++ containers isn't permitted, they would be passed as opaque pointers and C++ code could do nothing with them
* Using C++ types in Verona containers is less of a problem, but could also be restricted in the first iteration

Reply:
I'm not sure what you mean by 'container' here, since that isn't a language construct, but I think I more-or-less agree:

 - If you pass a Verona object to a foreign library, the foreign code sees a thing that looks like a pointer but that it can't actually dereference (in the non-CHERI case, this is actually an index into a table [possibly encoded in such a way that if you look at the bits it's a plausible pointer so that it doesn't clash with things that do tricks like use numbers < 64 KiB as special values] that contains a pointer to the Verona object, in the CHERI case it's probably a sealed capability to the object).  It has no methods.  You can store it in a `std::vector<void*>` or whatever, but it isn't actually very useful.
 - A C++ pointer looks in Verona like an external reference.  You can store it in a collection only if that collection can handle external references (doing anything with an external reference other than [perhaps] identity comparison requires presenting the region that contains the object).  Any method you call on it will (logically, at least) be an RPC.

Reply:
At least in the absence of CHERI, I don't really see any reason for pointers to Verona objects to show up as anything other than `void*` on the C++ side.  I'll add some explanation there.  

`Array[C++Type]` is a bit more interesting because I *can* imagine wanting to have `Array[SomeSandbox.int]`, for example.  If `C++Type` is a C++ class then it must be created in a sandboxed region, but `Array` can't be (probably - we might want to allow `Array` to be created in sandboxed regions so that we have a consistent class for 'contiguous blob of data', but I think creating a C array and doing an explicit copy probably makes more sense), so you can't actually express that type (or, rather, you can, but you can only use it to create a 0-element array).  You can express `Array[external_ref & C++Type` (or whatever the syntax is for external references), but this will then be a pair of a pointer to the sandbox and a pointer to the object and needs explicitly converting to a pointer to the object before it can be used, so there's a hint in the language that you don't want to be doing this.

Reply:
Exactly. Both cases of mixing lead to complicated semantics which I think we should avoid. Not so much for the compiler, but for the cognitive load on developers.

`Array[Sandbox.int]` is an interesting case, but since that's an alias to a Verona type, it's trivial. 

What I meant by `C++Type` was C++ classes, and as you say, ends up being an external reference, which needs not only pointer conversion but RPC-calling after that.